File: c_ext/dump_codepoints.c
#include <assert.h>
#include <stdio.h>

#include "lib/unicode.h"
#include "lib/util.h"

void help(const char* name_of_program, FILE* out_f) {
    fprintf(out_f,
        "\n"
        "    Usage:\n"
        "    * %s (input file)\n"
        "    * %s (input file) (output file)\n"
        "\n"
        "Dumps the UTF-8 code points of the input file to the output file,\n"
        "one human-readable integer per line.\n"
        "\n"
        "If an output file is not provided, writes to stdout.\n",
        name_of_program, name_of_program);
}

bool dump_code_points(const char* in_f_name, FILE* out) {
    if (!out) {
        return false;
    }

    char* str = NULL;
    size_t str_size = ~0ul;
    if (!util_read_file(in_f_name, &str, &str_size)) {
        return false;
    }

    unicode_code_point_t c;
    size_t str_index = 0;
    while (read_next_utf8(str, str_size, &str_index, &c)) {
        fprintf(out, "%u\n", c);
    }
    return true;
}

int main(int argc, char* argv[]) {
    const char* in_f;
    FILE* out_f;
    if (argc == 2) {
        in_f = argv[1];
        out_f = stdout;
    } else if (argc == 3) {
        in_f = argv[1];
        out_f = fopen(argv[2], "wb");
    } else {
        help(argv[0], stderr);
        return 1;
    }

    dump_code_points(in_f, out_f);
}

File: tools/dump_codepoints.py
def dump_help(name_of_program):
    print 'Usage:'
    print '* %s (input file)' % name_of_program
    print '* %s (input file) (output file)' % name_of_program
    print
    print 'Dumps the integer code points from (input file) to (output file), '
    print 'one per line.  If output file is not provided, uses stdout.'


def dump_codepoints(in_f_name, out_f_name):
    if out_f_name:
        out_f = open(out_f_name, 'wb')
    else:
        from sys import stdout
        out_f = stdout
    text = open(in_f_name).read().decode('utf-8')
    for c in text:
        out_f.write('%d\n' % ord(c))
    out_f.close()


def main():
    from sys import argv

    if len(argv) == 2:
        _, in_f = argv
        out_f = None
    elif len(argv) == 3:
        _, in_f, out_f = argv
    else:
        dump_help(argv[0])
        return

    dump_codepoints(in_f, out_f)


if __name__ == '__main__':
    main()
